paddle/fluid/framework/ir/mkldnn/fc_mkldnn_pass.cc

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/framework/ir/mkldnn/fc_mkldnn_pass.h"
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+#include "paddle/fluid/framework/eigen.h"
+#include "paddle/fluid/framework/lod_tensor.h"
+#include "paddle/fluid/platform/enforce.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+void Transpose(float* data, size_t rows, unsigned cols) {
+  std::vector<float> temp(rows * cols);
+
+#pragma omp parallel for collapse(2)
+  for (int i = 0; i < rows; i++) {
+    for (int j = 0; j < cols; j++) {
+      temp[j * rows + i] = data[i * cols + j];
+    }
+  }
+
+  std::copy(temp.begin(), temp.end(), data);
+}
+
+void FCMKLDNNPass::ApplyImpl(ir::Graph* graph) const {
```

Comment:
Since we have `fc_fuse_pass` and `mkldnn_placement_pass`, why we need `fc_mkldnn_pass`? 
After `fc_fuse_pass` and `mkldnn_placement_pass`, we could call mkldnn kernel of `fc_op`.

Reply:
It's because the weights of the layer need to be transposed in the fc mkldnn pass. This allows the mkl-dnn's algorithm to execute much more efficiently. I can further explain why is that if necessary.

Reply:
OK, this makes sense, but cloud we reuse some codes, seems only weight need reorder?

But another question, does this only for inference? If works on training, does it would cause some grad update issue? or print weight issue since format may be not nchw.

Reply:
What code shall be re-used in this case? The pass is taking care of checking whether the Input has correct dimensions and applies the transpose only in this case.

This op is only designed to work for inference.

Reply:
fc_mkldnn_pass now turns MKL-DNN's fully connected layer on if input dimensions are 2 or 4.